### PR TITLE
CaliForest - kobeguo2

### PR DESCRIPTION
**Pull request by KobeGuo99**
**Contributor:** Kobe Guo (kobeguo2)

**Contribution Type:** Model

**Description:**
This PR implements CaliForest, a calibrated Random Forest model for healthcare prediction tasks. The model uses out-of-bag (OOB) predictions and tree-level variance to perform post-hoc probability calibration via isotonic regression or logistic regression, improving probability reliability.

The implementation integrates with PyHealth’s BaseModel API and supports PyHealth training and evaluation workflows.

**Files to Review:**
- `pyhealth/models/califorest.py` — core model implementation
- `tests/core/test_califorest.py` — unit tests using synthetic data
- `examples/mimic4_califorest.py` — real-data ablation on MIMIC-IV (RF vs CaliForest, isotonic vs logistic)
- `docs/api/models/pyhealth.models.califorest.rst` — API documentation
- `docs/api/models.rst` — updated model index
- `pyhealth/models/__init__.py` — model registration

**Testing:**
- All unit tests pass locally (`pytest tests/core/test_califorest.py`)
- Example ablation results (MIMIC-IV subset):
  - RF AUROC: ~0.80
  - CaliForest (isotonic): ~0.81
  - CaliForest (logistic): ~0.80

**Notes:**
- Dataset path is configurable via `MIMIC4_ROOT` environment variable
- Synthetic data is used in tests to ensure fast CI execution

### CODE REVIEW

**Comment by joshuasteier** [comment]
Hi @KobeGuo99 , thanks for putting this together. The BaseModel integration, SampleDataset handling, the docs entry, and the model registration in __init__.py all look correct and match the pattern used by LogisticRegression in the package. 

The MIMIC-IV ablation in examples/mimic4_califorest.py is a nice touch and gives a real healthcare-relevant comparison.I have one thing I would like you to think through, and a few smaller suggestions.

I have one thing I would like you to think through, and a few smaller suggestions.
The main thing is around when the random forest actually gets trained. In forward:

if not self.is_fitted:
    self.fit_model(**kwargs)
    
    
This fits the RF and the calibrator on the first batch seen during forward, then freezes them for the rest of training. If someone uses the standard pyhealth.trainer.Trainer with normal mini-batching, the RF ends up trained on just that first batch and never sees the rest of the data. The test_model_backward test still passes because self.logit_scale has a gradient, but the forest itself is not being updated across epochs.Your MIMIC-IV example works around this by using batch_size=len(train_dataset), which feeds the whole training set in one batch. That is fine for the example, but it would be good to make this behavior explicit so future users do not hit it by surprise. A few ways you could handle it:
Add a clear note in the class docstring and in fit_model that CaliForest expects the full training set in one batch, and add an assertion or warning when it looks like mini-batching is happening.
Add a separate fit(train_dataset) method that materializes the full dataset once and fits the RF there. Then forward only does inference. This would decouple the RF from the torch training loop and make the behavior easier to reason about.
Keep forward as is but add a short section to the class docstring explaining the design choice and pointing users to the example.


Option 2 is probably the cleanest fit for how tree models usually work, but option 1 or 3 is much less work if you are short on time .Smaller suggestions:
The tests verify shapes and that a gradient flows, but they do not check the paper's main claim that CaliForest improves calibration over a vanilla random forest. It would strengthen the PR to add one test that compares Brier score between CaliForest and a plain RandomForestClassifier on synthetic data with maybe 200 to 500 samples, and asserts CaliForest is not worse. Also a quick test that isotonic and logistic produce different outputs on the same input would confirm both code paths are live.



The PR description mentions tree-level variance weighting, and I can see it in fit_model as weights = 1.0 / (variances + 1e-6). A short comment in the code pointing to the paper section that motivates this, and maybe a test that confirms the weighting changes the calibrator fit compared to unweighted, would make this feature easier for reviewers and users to verify.

predict_proba_numpy uses [:, 1] indexing which assumes binary classification. Could you add an assertion in fit_model that the labels are binary, or mention the binary-only scope in the class docstring? That way users with a multiclass task get a clear error instead of a silent wrong answer.

The unit tests use 4 samples with n_estimators=10. That is on the small side for OOB to be meaningful. Bumping to 50 or 100 synthetic samples would make the tests more representative.

One small thing: there is a parallel PR at #995 from a different team that also implements CaliForest. If you two want to compare notes, their implementation includes a min_oob_trees reliability threshold that is worth looking at for a possible follow-up addition. No action needed here, just a heads up.
Overall the structure is solid and the integration with PyHealth is done right. Happy to iterate on any of the above.

**Comment by KobeGuo99** [comment]
Thank you! This was a good catch. I updated CaliForest so it now requires an explicit `fit(train_loader)` call and no longer trains inside `forward()`. `forward()` now raises a clear error if called before fitting. I also updated the MIMIC-IV example to call `model.fit(train_loader)` explicitly before inference, added a binary-only note in the class docstring, and updated the tests accordingly. I also added a small test showing that the isotonic and logistic calibration paths produce different outputs. I really appreciate the feedback!

**Comment by joshuasteier** [comment]
Thanks @KobeGuo99 , this is a clean response to the feedback. The explicit `fit(train_loader)` method decouples the RF fitting from the torch training loop, the `forward` error message when called before fit is clear, the binary-label check in `fit_model` gives users a clear failure mode, and `test_isotonic_and_logistic_differ` confirms both calibration paths are active. The docstring note about binary scope and the fit-before-inference contract is exactly what I wanted.

A few small things worth thinking about, none blocking:

1. `self.logit_scale = nn.Parameter(torch.tensor(1.0))` is still defined and multiplied into the logits in `forward`. Under the old design where `forward` also did the fit, `logit_scale` gave `Trainer.fit()` something to update. Under the new design where training happens in `model.fit(train_loader)` and `forward` is inference-only, `logit_scale` sits as an unused trainable parameter. You can either remove it (simplest), or keep it and wire up a small optimiser step inside `fit(train_loader)` that tunes it on the train set (a Platt-scaling-on-top-of-calibration step). Either is fine; I would just not leave it dangling.

2. The unit tests still use 4 samples with `n_estimators=10`. OOB with so few samples is noisy. Bumping to 50 or 100 synthetic samples would make the tests more representative, and would also enable a quick Brier-score comparison test between CaliForest and a plain `RandomForestClassifier` to verify the paper's calibration claim. Not required for this PR, but worth it as a follow-up.

3. The comment on variance weighting (`CaliForest uses inverse tree-level variance so more stable predictions have greater influence during calibrator fitting`) explains the why, which is what I wanted. If you have the paper handy, a one-liner pointer to the equation or section would make it even easier for future reviewers to verify.

Happy to approve once you either remove or wire up `logit_scale`. The test additions are nice-to-have and can land in a follow-up PR.

**Comment by KobeGuo99** [comment]
Thank you! I removed the logit_scale from califorest.py and reran the CaliForest tests and the MIMIC-IV example after the change.

**Review by joshuasteier — APPROVED** [review]
Thanks @KobeGuo99.

One tiny nit, not blocking: `import torch.nn as nn` at the top of the file is now unused since `nn.Parameter` was the only thing using it. Feel free to remove the import on your next push, or leave it and a follow-up cleanup will catch it.

Approving. LGTM.
